Sources/PackageManagerDocs/Documentation.docc/Plugins/WritingBuildToolPlugin.md

```diff
@@ -0,0 +1,250 @@
+# Writing a build tool plugin
+
+@Metadata {
+    @Available("Swift", introduced: "5.6")
+}
+
+Create a build tool to process or generate files.
+
+## Overview
+
+The first step when writing a package plugin is to decide what kind of plugin you need.
+Implement a build tool plugin to generate source files that should be part of a build, or to perform other actions at the start of every build.
+Build tool plugins are invoked before a package is built in order to construct command invocations to run as part of the build.
+
+A build tool plugin can provide two kinds of commands:
+
+- term prebuild commands: commands that package manager runs before the build starts. Prebuild commands can generate an arbitrary number of output files with names that can't be predicted before running the command.
+- term build commands: commands that package manager incorporates into the build system's dependency graph and runs at the appropriate time during the build based on the existence and timestamps of their predefined inputs and outputs.
+
+> Note: If your goal is to provide an actions that you can perform at any time and are not associated with a build, implement a command plugin.
+> See <doc:WritingCommandPlugin> for details about creating a command plugin.
+
+With both prebuild and build commands, it is important to note that the build tool plugin doesn't do the work, rather it constructs the commands that the build runs later, and it is those commands that perform the work.
+The plugin can be quite small, and is often concerned with forming the command line for the build command that does the actual work.
+
+A build tool plugin is available to the package that defines it, and if there is a corresponding plugin product, it is also available to any other package that has a direct dependency on the package.
+
+### Build commands
+
+Prefer to create a build command over a prebuild command when the paths of all of the inputs and outputs are known before the command runs.
+Build tool command are more efficient because they provide the build system the information needed to efficiently decide when the build should invoke them.
```

Comment:
```suggestion
Build tool commands are more efficient because they provide the build system the information needed to efficiently decide when the build should invoke them.
```